pages/vpub_0x10/1-edk2-arm.md

```diff
@@ -0,0 +1,263 @@
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK2, 950000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK3, 750000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK4, 950000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK5, 900000),
+    /* This is not configured in the M2's Linux device tree
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK6, 1100000), */
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK7, 2000000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK8, 3300000),
+    RK8XX_VOLTAGE_INIT(MASTER_BUCK10, 1800000),
+```
+
+---
+
+## Step 2 - PCIe (M.2) Initialization
+
+`1.` What the existing code tells us:
+
+  ```c
+  if (Segment == PCIE_SEGMENT_PCIE20L2) {
+      GpioPinSetDirection(3, GPIO_PIN_PD1, GPIO_PIN_OUTPUT);
+  }
+  ```
+
+- Only Reset signal → GPIO PD1
+- No Power‑Enable on this board
+
+`2.` Hardkernel schematic findings:
+
+- Reset → GPIO 1 PA7
+- Power‑Enable → GPIO 0 PC6
+
+---
+
+## Step 2 - PCIe (M.2) Initialization
+
+`3.` Final M.2-specific code:
+
+```c
+EFIAPI
+PcieIoInit(UINT32 Segment) {
+    if (Segment == PCIE_SEGMENT_PCIE20L2) {
+        GpioPinSetDirection(1, GPIO_PIN_PA7, GPIO_PIN_OUTPUT);
+        GpioPinSetDirection(0, GPIO_PIN_PC6, GPIO_PIN_OUTPUT);
+    }
+}
+
+VOID EFIAPI
+PciePowerEn(UINT32 Segment, BOOLEAN Enable) {
+    if (Segment == PCIE_SEGMENT_PCIE20L2)
+        GpioPinWrite(0, GPIO_PIN_PC6, Enable);
+}
+
+VOID EFIAPI
+PciePeReset(UINT32 Segment, BOOLEAN Enable) {
+    if (Segment == PCIE_SEGMENT_PCIE20L2)
+        GpioPinWrite(1, GPIO_PIN_PA7, !Enable);
+}
+```
+
+---
+
+## Step 3: Fan control
+
+- Fan wired to GPIO 1 PA2 (PWM0_M2)
+- Small fan → low period (50 µs) + 50 % duty → quiet at low speed
+- Code:
+
+  ```c
+  PWM_DATA pwm_data = {
+      .ControllerID = PWM_CONTROLLER0,
+      .ChannelID    = PWM_CHANNEL0,
+      .PeriodNs     = 50000,
+      .DutyNs       = 50000,
+      .Polarity     = FALSE,
+  };
+
+  VOID EFIAPI PwmFanIoSetup(VOID) {
+      GpioPinSetFunction(1, GPIO_PIN_PA2, 0xB);   // PWM0_M2
+      RkPwmSetConfig(&pwm_data);
+      RkPwmEnable(&pwm_data);
+  }
+  ```
+
+---
+
+## Additional board-specific tweaks
+
+| FEATURE | CURRENT STATE | COMMENTS |
+| --- | --- | --- |
+| LEDs | Not functional | Need to map GPIOs |
+| USB 2.0 | Functional | |
+| USB 3.0 | Not functional | Works once booted into OS |
+| Booting Windows | Not functional | Need proper ACPI tables |
+
+---
+
+## Building the image
+
+- Root of edk2‑rk3588 → build.sh
+- Command:
+
+  ```bash
+  ./build.sh --device odroid-m2
+
+  => FIT build done
+  => Building 8MB NOR FLASH IMAGE
+  ...
+  Build done: RK3588_NOR_FLASH.img
+  ```
+
+---
+
+## Flashing to MicroSD
+
+- Odroid M2 boots from microSD (no onboard NAND on dev boards)
+- Flashing command:
+
+  ```bash
+  sudo dd if=RK3588_NOR_FLASH.img of=/dev/sdb bs=4M status=progress
+  ```
+
+---
+
+## Boot result
+
+- Insert SD card, power on
+- Image inserted → board powers up → UEFI splash appears
+- Ubuntu 25.04 boots (from FIT image)
+
+---
+theme: slidev-template/theme
+layout: cover
+background: /intro.png
+class: text-center
+---
```

Comment:
Looks like a mistake:
```suggestion
theme: slidev-template/theme
layout: cover
background: /intro.png
class: text-center
---
```

Reply:
What do you mean? This renders correctly for me